tensorboard/plugins/custom_scalar/tf_custom_scalar_dashboard/tf-custom-scalar-margin-chart-card.html

```diff
@@ -0,0 +1,579 @@
+          chart.resetDomain();
+        }
+      },
+      _csvUrl(nameToDataSeries, dataSeriesName) {
+        const baseUrl = this._downloadDataUrl(nameToDataSeries, dataSeriesName);
+        return tf_backend.addParams(baseUrl, {format: 'csv'});
+      },
+      _jsonUrl(nameToDataSeries, dataSeriesName) {
+        const baseUrl = this._downloadDataUrl(nameToDataSeries, dataSeriesName);
+        return tf_backend.addParams(baseUrl, {format: 'json'});
+      },
+      _downloadDataUrl(nameToDataSeries, dataSeriesName) {
+        const dataSeries = nameToDataSeries[dataSeriesName];
+        const getVars = {
+            tag: dataSeries.getTag(),
+            run: dataSeries.getRun(),
+        };
+        return tf_backend.addParams(
+              tf_backend.getRouter().pluginRoute(
+                  'custom_scalars', '/download_data'), getVars);
+      },
+      _computeXComponentsCreationMethod(xType) {
+        return () => vz_line_chart.getXComponents(xType);
+      },
+      _createProcessDataFunction(marginChartSeries) {
+        // This function is called when data is received from the backend.
+        return ((scalarChart, run, data) => {
+          if (data.regex_valid) {
+            // The user's regular expression was valid.
+            // Incorporate these newly loaded values.
+            const newMapping = _.clone(this._nameToDataSeries);
+            const tagsNotFound = [];
+            _.forEach(marginChartSeries, tagsObject => {
+              let tagNotFound = false;
+              const scalarEvents = data.tag_to_events[tagsObject.value];
+              const lowerBounds = data.tag_to_events[tagsObject.lower];
+              const upperBounds = data.tag_to_events[tagsObject.upper];
+
+              if (_.isUndefined(scalarEvents)) {
+                tagsNotFound.push(tagsObject.value);
+                tagNotFound = true;
+              }
+              if (_.isUndefined(lowerBounds)) {
+                tagsNotFound.push(tagsObject.lower);
+                tagNotFound = true;
+              }
+              if (_.isUndefined(upperBounds)) {
+                tagsNotFound.push(tagsObject.upper);
+                tagNotFound = true;
+              }
+
+              if (tagNotFound) {
+                return;
+              }
+
+              const dataPoints = scalarEvents.map(datum => ({
+                wall_time: new Date(datum[0] * 1000),
+                step: datum[1],
+                scalar: datum[2],
+              }));
+              
+              // Associate the lower bounds with the correct data points.
+              for (let i = j = 0;
+                   i < dataPoints.length && j < lowerBounds.length;) {
+                const boundDataPoint = lowerBounds[j];
+                const step = boundDataPoint[1];
+                const bound = boundDataPoint[2];
+                if (step < dataPoints[i].step) {
+                  j++;
+                } else if (step > dataPoints[i].step) {
```

Comment:
This condition can only be true if we've "missed" a lower bound for a given point in `dataPoints`.  By advancing past `dataPoints[i]` without assigning anything to its `lower` property we're setting things up for an error down the line when we compute `fillArea`, right?  Ditto for the upper bound logic below.

There are various ways we could attempt to handle those gaps (setting the missing bound to `scalar` or omitting that data point entirely) but I think it would be perhaps least surprising to just emit an error and refuse to plot a series if we can't find matching-step lower and upper bounds for each data point.

If we do that, I think there might be a better way to structure this algorithm so that it basically just iterates over the data points and for each one, scans ahead in both bound arrays until it finds bounds at matching steps (and if it doesn't find them, it reports an error).  Alternatively we could be even stricter and not permit more bounds than data points, either (while that case would not force a plotting error, it's unlikely to come up in practice and might still represent incorrect usage).  That would make the algorithm super simple since you'd just walk all three lists in sync and assert that the steps match.